geo/src/algorithm/centroid.rs

```diff
@@ -340,28 +339,136 @@ where
+                    let mut centroid_accum = centroid_1d_accum.unwrap_or_default();
+
+                    fn length<T>(geometry: &Geometry<T>) -> T
+                    where
+                        T: CoordFloat + FromPrimitive + Sum + Default,
+                    {
+                        match geometry {
+                            Geometry::Line(l) => l.euclidean_length(),
+                            Geometry::LineString(l) => l.euclidean_length(),
+                            Geometry::MultiLineString(l) => l.euclidean_length(),
+                            Geometry::GeometryCollection(geometry_collection) => {
+                                geometry_collection.iter().map(length).sum()
+                            }
+                            Geometry::Point(_)
+                            | Geometry::MultiPoint(_)
+                            | Geometry::Polygon(_)
+                            | Geometry::MultiPolygon(_)
+                            | Geometry::Rect(_)
+                            | Geometry::Triangle(_) => T::zero(),
+                        }
+                    }
+                    let weight = length(geometry);
+
+                    if weight == T::zero() {
+                        continue;
+                    }
+
+                    centroid_accum.0 = centroid_accum.0 + centroid * weight;
+                    centroid_accum.1 = centroid_accum.1 + weight;
+                    centroid_1d_accum = Some(centroid_accum);
+                }
+                (Dimensions::TwoDimensional, Some(centroid)) => {
+                    let mut centroid_accum = centroid_2d_accum.unwrap_or_default();
+
+                    // REVIEW: unsigned?
+                    let weight = geometry.unsigned_area();
+
+                    if weight == T::zero() {
+                        continue;
+                    }
+
+                    centroid_accum.0 = centroid_accum.0 + centroid * weight;
+                    centroid_accum.1 = centroid_accum.1 + weight;
+                    centroid_2d_accum = Some(centroid_accum);
+                }
+            };
         }
-        let sum = self.iter().fold(
-            Point::new(T::zero(), T::zero()),
-            |a: Point<T>, b: &Point<T>| Point::new(a.x() + b.x(), a.y() + b.y()),
-        );
-        Some(Point::new(
-            sum.x() / T::from(self.0.len()).unwrap(),
-            sum.y() / T::from(self.0.len()).unwrap(),
-        ))
+
+        centroid_2d_accum
+            .or(centroid_1d_accum)
+            .or(centroid_0d_accum)
+            .map(|(centroid, weight)| centroid / weight)
+    }
+}
+
+impl<T> Centroid for Triangle<T>
+where
+    T: CoordFloat + FromPrimitive + Sum,
+{
+    type Output = Point<T>;
+    fn centroid(&self) -> Self::Output {
+        let coord = self.0 + self.1 + self.2 / T::from_usize(3).unwrap();
+        Point::from(coord)
+    }
+}
+
+fn centroid_of_coords<T>(
+    mut coords_iter: impl Iterator<Item = Coordinate<T>>,
```

Comment:
I considered implementing this as `impl Centroid for CoordsIter` but thought it might be a footgun, since, e.g. Polygon implements CoordsIter, but we'd never want to use the `impl CoordsIter for Centroid` for a `Polygon`.